File: snippets/formulae/pi-extraction.py
import math

# Inputs - these are set from the API results.
# Note that all times are in seconds.
planet_dict = {}
planet_dict["total_cycles"] = (
    30  # End time in seconds - start time in seconds / cycle time
)
planet_dict["cycle_time"] = 30 * 60  # 30 minutes, value from API cycleTime * 60
planet_dict["qty_per_cycle"] = 6965

# These constants are the defaults in dgmAttributeTypes. They may change.
decay_factor = 0.012  # Dogma attribute 1683 for this pin typeID
noiseFactor = 0.8  # Dogma attribute 1687 for this pin typeID

# Populated by data from the PI ESI Endpoint

cycle_time = planet_dict["cycle_time"]
planet_dict["total_qty"] = 0
quantityPerCycle = planet_dict["qty_per_cycle"]
barWidth = cycle_time / 900
total_cycles = planet_dict["total_cycles"]
for cycle in range(0, int(total_cycles)):
    t = (cycle + 0.5) * barWidth
    decay_value = quantityPerCycle / (1 + t * decay_factor)
    phaseShift = pow(quantityPerCycle, 0.7)

    sinA = math.cos(phaseShift + t * (1 / 12))
    sinB = math.cos(phaseShift / 2 + t * 0.2)
    sinC = math.cos(t * 0.5)

    sinStuff = (sinA + sinB + sinC) / 3
    sinStuff = max(sinStuff, 0)

    barHeight = decay_value * (1 + noiseFactor * sinStuff)

    planet_dict["total_qty"] += barWidth * barHeight

planet_dict["average_qty"] = planet_dict["total_qty"] / planet_dict["total_cycles"]

File: snippets/formulae/pi-extraction.cs
int[] CalculateExtractorValues() {

    //Inputs - these are set from the API results.
    //Note that all times are in seconds.
    int duration = 171000; //1d 23h 30m; from API expiryTime-installTime
    int cycleTime = 30 * 60; //30 minutes, value from API cycleTime * 60
    int quantityPerCycle = 6965;

    //These constants are the defaults in dgmAttributeTypes. They may change.   
    const float decayFactor = 0.012f; //Dogma attribute 1683 for this pin typeID
    const float noiseFactor = 0.8f;   //Dogma attribute 1687 for this pin typeID
           
    int numIterations = duration / cycleTime;
    float barWidth = cycleTime / 900f;
    int[] values = new int[numIterations];

    for (int i = 0; i < numIterations; i++) {
        float t = (i + 0.5f)*barWidth;
        float decayValue = quantityPerCycle/(1 + t * decayFactor);
        double phaseShift = Math.Pow(quantityPerCycle, 0.7f);
        
        double sinA = Math.Cos(phaseShift + t * (1/12f));
        double sinB = Math.Cos(phaseShift / 2 + t * 0.2f);
        double sinC = Math.Cos(t * 0.5f);
   
        double sinStuff = Math.Max((sinA + sinB + sinC) / 3, 0);
   
        double barHeight = decayValue * (1 + noiseFactor * sinStuff);
        values[i] = (int) (barWidth * barHeight);
   }
   return values;
}
